src/debug.cc

```diff
@@ -91,6 +416,209 @@ ResyncDebugLog(FILE *newFile)
+            log(message.header, message.body);
+    }
+    const auto writtenNow = written - writtenEarlier;
+
+    if (const auto totalCount = ourCount + theirCount) {
+        debugs(0, 5, "wrote " << writtenNow << " out of " << totalCount << '=' <<
+               ourCount << '+' << theirCount << " early messages to " << name);
+    }
+}
+
+bool
+DebugChannel::saveMessage(const DebugMessageHeader &header, const std::string &body)
+{
+    if (!earlyMessages)
+        return false; // we have stopped saving early messages
+
+    if (header.level > EarlyMessagesLevel)
+        return false; // this message is not important enough to save
+
+    // Given small EarlyMessagesLevel, only a Squid bug can cause so many
+    // earlyMessages. Saving/dumping excessive messages correctly is not only
+    // difficult but is more likely to complicate triage than help: It is the
+    // first earlyMessages that are going to be the most valuable. Our assert()
+    // will dump them if at all possible.
+    assert(earlyMessages->size() < 1000);
+
+    earlyMessages->emplace_back(header, body);
+    return true;
+}
+
+void
+DebugChannel::writeToStream(FILE &destination, const DebugMessageHeader &header, const std::string &body)
+{
+    fprintf(&destination, "role=%s # %s%s| %s\n",
+            header.role_XXX,
+            debugLogTime(header.timestamp),
+            debugLogKid(),
+            body.c_str());
+    noteWritten(header);
+}
+
+void
+DebugChannel::noteWritten(const DebugMessageHeader &header)
+{
+    ++written;
+    lastWrittenRecordNumber = header.recordNumber;
+}
+
+/* CacheLogChannel */
+
+void
+CacheLogChannel::log(const DebugMessageHeader &header, const std::string &body)
+{
+    if (header.recordNumber <= lastWrittenRecordNumber)
+        return;
+
+    if (earlyMessages)
+        return (void)saveMessage(header, body);
+
+    if (!TheLog.file())
+        return;
+
+    writeToStream(*TheLog.file(), header, body);
+    fflush(TheLog.file());
+}
+
+/* StderrChannel */
+
+bool
+StderrChannel::shouldWrite(const int level) const
+{
+    if (!stderr)
+        return false; // nowhere to write
+
+    if (ExplicitStderrLevelSet) // explicit admin restrictions (-d)
+        return level <= ExplicitStderrLevel;
+
+    // whether the given level is allowed by emergency handling circumstances
+    // (coveringForCacheLog) or configuration aspects (e.g., -k or -z)
+    return coveringForCacheLog || level <= DefaultStderrLevel;
+}
+
+void
+StderrChannel::log(const DebugMessageHeader &header, const std::string &body)
+{
+    if (header.recordNumber <= lastWrittenRecordNumber)
+        return;
+
+    if (saveMessage(header, body))
+        return;
+
+    if (!shouldWrite(header.level))
+        return;
+
+    // We must write this eligible unsaved message, but we must log previously
+    // saved early messages before writeToStream() below to avoid reordering.
+    stopEarlyMessageCollection();
```

Comment:
While testing, I noticed that each time we log a buffered early message, this method is vainly called (and so are nested logSaved/logSavedAnd()). I guess that we can optimize a little, adding a condition:
```suggestion
if (earlyMessages)
    stopEarlyMessageCollection();
```

Reply:
Done.